### PR TITLE
feat: system tests use new deployers and lender contracts

### PR DESCRIPTION
**Pull request by manuelpolzhofer**
**Deployer: Lender, Borrower**
- Deployer using auth pattern
- the user who created the deployer doesn't have access after deployment is done
- added permissions to deployer

**RootAdmin Contract**
- after borrower and lender are deployed only rootAdmin has access.
- root admin completes the deployment 
- root admin can give access to contracts

**Feat**
- pricePool Fab
- pricePool added to borrower deployer
- test_utils.sol
  - has methods to deploy all Tinlake contracts
  - handles the user and admin contract

**Fix**
- Assessor `calcTokenPrice` uses tranche parameter

### CODE REVIEW

File: src/core/deployer.sol [hunk 1]
```diff
@@ -12,187 +11,68 @@
...
-        principal = new Principal();
-        principal.rely(msg.sender);
-        principal.deny(address(this));
-    }
-}
+        // Borrower  Depends
+        borrowerDeployer.collector().depend("distributor", distributor_);
+        // todo needs to be tranche after mock
+        borrowerDeployer.shelf().depend("lender", distributor_);
+        borrowerDeployer.collector().depend("distributor", distributor_);
```
Comment (ilinzweilin) [inline src/core/deployer.sol]: dupl line 🙈

File: src/core/deployer.sol [hunk 2]
```diff
@@ -12,187 +11,68 @@
...
-        title = titlefab.newTitle(name, symbol);
-        title.rely(god);
-    }
-
-    function deployLightSwitch() public {
-        lightswitch = lightswitchfab.newLightSwitch();
-        lightswitch.rely(god);
-    }
+        // distributor allowed to call
+        borrowerDeployer.shelf().rely(address(lenderDeployer.distributor()));
```
Comment (ilinzweilin) [inline src/core/deployer.sol]: don't think that this is required. distributor just calls public methods from shelf

File: src/core/deployer.sol [hunk 3]
```diff
@@ -12,187 +11,68 @@
...
 
-contract CollectorFab {
-    function newCollector(address distributor, address shelf, address pile, address threshold) public returns (Collector collector) {
-        collector = new Collector(distributor, shelf, pile, threshold);
-        collector.rely(msg.sender);
-        collector.deny(address(this));
-    }
-}
+    function wireDepends() public auth {
+        address distributor_ = address(lenderDeployer.distributor());
```
Comment (ilinzweilin) [inline src/core/deployer.sol]: maybe we also need to make sure here that the addresses already exist

File: src/core/deployer.sol [hunk 4]
```diff
@@ -12,187 +11,68 @@
...
-        shelf.rely(god);
     }
 
-    // note: this method will be refactored with the new lender side contracts, we will rely on God once more
-    //and the Pile should articulate that it depends on the distributor, not a generic "lender".
-    function deployDistributor(address currency_) public {
-        distributor = distributorFab.newDistributor(address(shelf), currency_);
-        shelf.depend("lender", address(distributor));
+    function denyBorrowAdmin(address usr) public auth {
+        borrowerDeployer.principal().deny(usr);
```
Comment (ilinzweilin) [inline src/core/deployer.sol]: what about threshold and collector

File: src/core/deployer.sol [hunk 5]
```diff
@@ -12,187 +11,68 @@
...
+        borrowerDeployer.pile().deny(usr);
     }
 
-    function deployPrincipal() public {
-        principal = principalFab.newPrincipal();
-        principal.rely(god);
+    function relyBorrowAdmin(address usr) public auth {
+        borrowerDeployer.principal().rely(usr);
+        borrowerDeployer.shelf().rely(usr);
+        borrowerDeployer.pile().rely(usr);
```
Comment (ilinzweilin) [inline src/core/deployer.sol]: same

File: src/core/deployer.sol [hunk 6]
```diff
@@ -12,187 +11,68 @@
...
-        address distributor_ = address(distributor);
-        address shelf_ = address(shelf);
-        address collector_ = address(collector);
+    // todo: currently only needed for testing.
+    // test wants to create a specific setup
+    function relyRootAdmin(address usr) public auth {
+        borrowerDeployer.title().rely(usr);
+        borrowerDeployer.principal().rely(usr);
+        borrowerDeployer.shelf().rely(usr);
+        borrowerDeployer.pile().rely(usr);
```
Comment (ilinzweilin) [inline src/core/deployer.sol]: same

File: src/core/borrower/deployer.sol [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    function deployShelf(address currency_) public auth {
+        shelf = shelffab.newShelf(currency_, address(title), address(pile), address(principal));
+        shelf.rely(mainDeployer);
+    }
+
+    function deployPrincipal() public auth {
+        principal = principalFab.newPrincipal();
+        principal.rely(mainDeployer);
+    }
+
```
Comment (ilinzweilin) [inline src/core/borrower/deployer.sol]: I think we are also missing the CreditLine fab here

File: src/core/deployer.sol (hunk 4 above)
Comment (manuelpolzhofer) [reply src/core/deployer.sol]: yes, I only added the once which are currently need to fix the tests.

I will add the rest

File: src/core/deployer.sol (hunk 3 above)
Comment (manuelpolzhofer) [reply src/core/deployer.sol]: yes. I will add more safety checks. It is only called once so it makes sense

File: src/core/borrower/deployer.sol (hunk 7 above)
Comment (manuelpolzhofer) [reply src/core/borrower/deployer.sol]: I think we should rename the `fabs`.

PrincipalFab and CreditLineFab should both have a method called `newCeiling`

In the deployer i would use an abstract contract called `CeilingFabLike`

File: src/core/lender/assessor.sol [hunk 8]
```diff
@@ -78,13 +78,13 @@ contract Assessor is Math, DSNote, Auth {
         return _calcSeniorAssetValue(poolValue, trancheReserve, SeniorTrancheLike(tranche).debt(), _juniorReserve());
     }
 
-    function calcTokenPrice() public returns (uint) {
-        return mul(_calcTokenPrice(), tokenAmountForONE);
+    function calcTokenPrice(address tranche) public returns (uint) {
```
Comment (manuelpolzhofer) [inline src/core/lender/assessor.sol]: The method is called by the operator and not the tranche.

That's the reason why `msg.sender` is not working.

File: src/core/test/system/system.t.sol [hunk 9]
```diff
@@ -15,194 +15,93 @@
```
Comment (manuelpolzhofer) [inline src/core/test/system/system.t.sol]: The current systems are basically testing the borrower modules.

A method called `setupCurrencyOnLender` guarantees that the lender has currency for the tests

**Pull request merged**